protocols/request-response/src/handler.rs

```diff
@@ -191,9 +195,15 @@ where
     }
 }
 
-/// The events emitted by the [`RequestResponseHandler`].
+#[deprecated(
+    since = "0.23.0",
+    note = "Use re-exports that omit `RequestResponse` prefix, i.e. `libp2p::request_response::handler::HandlerEvent`"
+)]
+pub type RequestResponseHandlerEvent<TCodec> = HandlerEvent<TCodec>;
+
+/// The events emitted by the [`Handler`].
 #[doc(hidden)]
-pub enum RequestResponseHandlerEvent<TCodec>
+pub enum HandlerEvent<TCodec>
```

Comment:
Would just `Event` be pushing it too far? :D

Reply:
In case it isn't I think the `handler` module should be private so this `Event` can clash with the other one.

Reply:
Intuitively `libp2p::request_response::handler::Event` reads better than `libp2p::request_response::handler::HandlerEvent` and would be inline with the convention this pull request is enforcing.

Reply:
yeah makes sense to me, thanks for the suggestion,
So, there is `request_response::Event` and `request_response::handler::Event` now. Can you expand on `Event`'s clashing Thomas?  
There will still be `request_response::RequestResponseHandlerEvent` which is deprecated and suggests `request_response::handler::Event`, if we make `handler` private how can we export both?

Reply:
Yeah if we make `handler` private, we can no longer export it. What I meant with clashing is that from the outside, there is really only one event that is interesting for users and it is the `NetworkBehaviour`'s `OutEvent`.

Given that we are already making this a breaking change, we might as well make `handler` private. In my opinion, no one should be depending on just the `ConnectionHandler`.

Reply:
Thanks for the explanation Thomas. `Handler` and `Event` can't be private because `NetworkBehaviour::ConnectionBehaviour` associated type needs to be public, and by consequence `ConnectionHandler`'s  also

Reply:
> Thanks for the explanation Thomas. `Handler` and `Event` can't be private because `NetworkBehaviour::ConnectionBehaviour` associated type needs to be public, and by consequence `ConnectionHandler`'s also

Hmm that is news to me! We have a lot of protocols where the `ConnectionHandler` implementation is completely private to the crate. For example `ping`: https://github.com/libp2p/rust-libp2p/blob/5755942d7742f9d0432c346318e25da504b65370/protocols/ping/src/lib.rs#L45

Reply:
I just tested it locally and things compile if I make the `handler` module private. However, I'd suggest that we don't do this for now. I think together with https://github.com/libp2p/rust-libp2p/pull/3159#discussion_r1041760501, we can make this completely non-breaking?

Reply:
Thanks for digging this deeper Thomas! Did you understand what is it that with `request-response` that was making the compiler complain?

Reply:
I think the issue is fairly subtle:

If you have a public type that implements a trait, the compiler enforces that the associated types also need to be public. It does however not check whether the type is within a public or private module in regards to the crate.

In our usecase, this means we can have a `pub struct Handler` that implements `ConnectionHandler` but define it within a crate-private module `handler`. This may or may not be a bug in Rust but I think one can argue that it makes sense:

The API contract promised to the user is that `Behaviour` implements `NetworkBehaviour`. `NetworkBehaviour` enforces that `NetworkBehaviour::ConnectionHandler` implements `IntoConnectionHandler`. The user can refer to this type as `<Behaviour::ConnectionHandler as IntoConnectionHandler>` and call all public APIs on that. They don't need to know the concrete type that implements the functionality. This way, only the minimal promised API is actually exposed.